ICE `invalid immediate for given destination place: scalar value has wrong size`

<!--
[31mICE[0m: Rustc ./BC3F301FE4B2C0AC50F0C3F13FAF1D2AE4AD05443021AFF4CB58B5260D844868.rs '-Zmir-enable-passes=+DataflowConstProp -ooutputfile -Zdump-mir-dir=dir' 'thread 'rustc' ($TID) panicked at /rustc-dev/5b150d238fbd4fe7bc2cd3140d8e6fb4406099fa/compiler/rustc_const_eval/src/interpret/operand.rs:123:17: 'assertion `left == right` failed: invalid immediate for given destination place: scalar value has wrong size'', 'thread 'rustc' ($TID) panicked at /rustc-dev/5b150d238fbd4fe7bc2cd3140d8e6fb4406099fa/compiler/rustc_const_eval/src/interpret/operand.rs:123:17: 'assertion `left == right` failed: invalid immediate for given destination place: scalar value has wrong size''
File: /tmp/im/BC3F301FE4B2C0AC50F0C3F13FAF1D2AE4AD05443021AFF4CB58B5260D844868.rs
-->

snippet
````rust
//@compile-flags: -Zmir-enable-passes=+DataflowConstProp --crate-type lib --edition=2021
pub async fn a() -> u32 {
    unsafe { std::mem::transmute(1usize) }
}

````

original:
````rust
//@ edition:2018
// .gitattributes file in this directory should enforce it.

/// [https://bloob.blob]
pub async fn a() -> u32 { std::mem::transmute(1usize) }

````

Version information
````
rustc 1.94.0-nightly (5b150d238 2025-12-11)
binary: rustc
commit-hash: 5b150d238fbd4fe7bc2cd3140d8e6fb4406099fa
commit-date: 2025-12-11
host: x86_64-unknown-linux-gnu
release: 1.94.0-nightly
LLVM version: 21.1.5

````



Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zmir-enable-passes=+DataflowConstProp --crate-type lib --edition=2021`

<details><summary><strong>Program output</strong></summary>
<p>

```
error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
 --> BC3F301FE4B2C0AC50F0C3F13FAF1D2AE4AD05443021AFF4CB58B5260D844868.rs:3:14
  |
3 |     unsafe { std::mem::transmute(1usize) }
  |              ^^^^^^^^^^^^^^^^^^^
  |
  = note: source type: `usize` (64 bits)
  = note: target type: `u32` (32 bits)


thread 'rustc' (1521152) panicked at /rustc-dev/c4dc70ee0ad7f811fb32e5bed9cd6c7b37beed4e/compiler/rustc_const_eval/src/interpret/operand.rs:123:17:
assertion `left == right` failed: invalid immediate for given destination place: scalar value has wrong size
  left: Size(8 bytes)
 right: Size(4 bytes)
stack backtrace:
   0:     0x7f09c44c6133 - <<std[4ba5a76045f2e409]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[f3ddf6de697c19ed]::fmt::Display>::fmt
   1:     0x7f09c4c10608 - core[f3ddf6de697c19ed]::fmt::write
   2:     0x7f09c44dc846 - <std[4ba5a76045f2e409]::sys::stdio::unix::Stderr as std[4ba5a76045f2e409]::io::Write>::write_fmt
   3:     0x7f09c449c8f8 - std[4ba5a76045f2e409]::panicking::default_hook::{closure#0}
   4:     0x7f09c44ba003 - std[4ba5a76045f2e409]::panicking::default_hook
   5:     0x7f09c350a247 - std[4ba5a76045f2e409]::panicking::update_hook::<alloc[7f145662aca8c9c]::boxed::Box<rustc_driver_impl[6aae568bf99046c3]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f09c44ba2e2 - std[4ba5a76045f2e409]::panicking::panic_with_hook
   7:     0x7f09c449c9b8 - std[4ba5a76045f2e409]::panicking::panic_handler::{closure#0}
   8:     0x7f09c44911d9 - std[4ba5a76045f2e409]::sys::backtrace::__rust_end_short_backtrace::<std[4ba5a76045f2e409]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f09c449e56d - __rustc[4d7f59a3bb196642]::rust_begin_unwind
  10:     0x7f09c10b853c - core[f3ddf6de697c19ed]::panicking::panic_fmt
  11:     0x7f09c2f22503 - core[f3ddf6de697c19ed]::panicking::assert_failed_inner
  12:     0x7f09c3abee2e - core[f3ddf6de697c19ed]::panicking::assert_failed::<rustc_abi[323bcc08463c2da1]::Size, rustc_abi[323bcc08463c2da1]::Size>
  13:     0x7f09c6969ae1 - <rustc_const_eval[a6b790c01f68a297]::interpret::operand::Immediate>::assert_matches_abi::<rustc_const_eval[a6b790c01f68a297]::interpret::eval_context::InterpCx<rustc_const_eval[a6b790c01f68a297]::const_eval::dummy_machine::DummyMachine>>.cold
  14:     0x7f09c57fcdfd - <rustc_const_eval[a6b790c01f68a297]::interpret::eval_context::InterpCx<rustc_const_eval[a6b790c01f68a297]::const_eval::dummy_machine::DummyMachine>>::write_immediate_to_mplace_no_validate
  15:     0x7f09c57fcd5b - <rustc_const_eval[a6b790c01f68a297]::interpret::eval_context::InterpCx<rustc_const_eval[a6b790c01f68a297]::const_eval::dummy_machine::DummyMachine>>::write_immediate_no_validate::<rustc_const_eval[a6b790c01f68a297]::interpret::place::PlaceTy>
  16:     0x7f09c3c82c97 - rustc_mir_transform[47693a1e9fa3d5e6]::dataflow_const_prop::try_write_constant
  17:     0x7f09c3c82ec2 - rustc_mir_transform[47693a1e9fa3d5e6]::dataflow_const_prop::try_write_constant
  18:     0x7f09c3c78062 - <rustc_mir_transform[47693a1e9fa3d5e6]::dataflow_const_prop::Collector>::try_make_constant
  19:     0x7f09c3cee0b8 - <rustc_mir_transform[47693a1e9fa3d5e6]::dataflow_const_prop::Collector as rustc_mir_dataflow[db5e4e02463b2c0b]::framework::visitor::ResultsVisitor<rustc_mir_transform[47693a1e9fa3d5e6]::dataflow_const_prop::ConstAnalysis>>::visit_after_primary_statement_effect
  20:     0x7f09c3c9e96a - <rustc_mir_transform[47693a1e9fa3d5e6]::dataflow_const_prop::DataflowConstProp as rustc_mir_transform[47693a1e9fa3d5e6]::pass_manager::MirPass>::run_pass
  21:     0x7f09c4c073ba - rustc_mir_transform[47693a1e9fa3d5e6]::run_optimization_passes
  22:     0x7f09c5a11152 - rustc_mir_transform[47693a1e9fa3d5e6]::optimized_mir
  23:     0x7f09c5a10d6f - rustc_query_impl[23ea27e8a90d8b37]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[23ea27e8a90d8b37]::query_impl::optimized_mir::dynamic_query::{closure#2}::{closure#0}, rustc_middle[21423b4d0c52388d]::query::erase::Erased<[u8; 8usize]>>
  24:     0x7f09c4c18806 - rustc_query_system[2e8d7520d0881f27]::query::plumbing::try_execute_query::<rustc_query_impl[23ea27e8a90d8b37]::DynamicConfig<rustc_query_system[2e8d7520d0881f27]::query::caches::DefIdCache<rustc_middle[21423b4d0c52388d]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[23ea27e8a90d8b37]::plumbing::QueryCtxt, false>
  25:     0x7f09c4c17ddb - rustc_query_impl[23ea27e8a90d8b37]::query_impl::optimized_mir::get_query_non_incr::__rust_end_short_backtrace
  26:     0x7f09c5fb96c1 - <rustc_middle[21423b4d0c52388d]::ty::context::TyCtxt>::coroutine_layout
  27:     0x7f09c554898f - rustc_ty_utils[1163db851dfce16c]::layout::layout_of_uncached
  28:     0x7f09c5538bc9 - rustc_ty_utils[1163db851dfce16c]::layout::layout_of
  29:     0x7f09c5538b2a - rustc_query_impl[23ea27e8a90d8b37]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[23ea27e8a90d8b37]::query_impl::layout_of::dynamic_query::{closure#2}::{closure#0}, rustc_middle[21423b4d0c52388d]::query::erase::Erased<[u8; 16usize]>>
  30:     0x7f09c55379e3 - rustc_query_system[2e8d7520d0881f27]::query::plumbing::try_execute_query::<rustc_query_impl[23ea27e8a90d8b37]::DynamicConfig<rustc_query_system[2e8d7520d0881f27]::query::caches::DefaultCache<rustc_middle[21423b4d0c52388d]::ty::PseudoCanonicalInput<rustc_middle[21423b4d0c52388d]::ty::Ty>, rustc_middle[21423b4d0c52388d]::query::erase::Erased<[u8; 16usize]>>, false, true, false>, rustc_query_impl[23ea27e8a90d8b37]::plumbing::QueryCtxt, false>
  31:     0x7f09c553761a - rustc_query_impl[23ea27e8a90d8b37]::query_impl::layout_of::get_query_non_incr::__rust_end_short_backtrace
  32:     0x7f09c5ce8a89 - <rustc_mir_transform[47693a1e9fa3d5e6]::known_panics_lint::KnownPanicsLint as rustc_mir_transform[47693a1e9fa3d5e6]::pass_manager::MirLint>::run_lint
  33:     0x7f09c4c0c59e - rustc_mir_transform[47693a1e9fa3d5e6]::run_analysis_to_runtime_passes
  34:     0x7f09c5813000 - rustc_mir_transform[47693a1e9fa3d5e6]::mir_drops_elaborated_and_const_checked
  35:     0x7f09c58129a9 - rustc_query_impl[23ea27e8a90d8b37]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[23ea27e8a90d8b37]::query_impl::mir_drops_elaborated_and_const_checked::dynamic_query::{closure#2}::{closure#0}, rustc_middle[21423b4d0c52388d]::query::erase::Erased<[u8; 8usize]>>
  36:     0x7f09c4d9a637 - rustc_query_system[2e8d7520d0881f27]::query::plumbing::try_execute_query::<rustc_query_impl[23ea27e8a90d8b37]::DynamicConfig<rustc_data_structures[e7ed3fe572b1bf09]::vec_cache::VecCache<rustc_span[75d45f3c27fef647]::def_id::LocalDefId, rustc_middle[21423b4d0c52388d]::query::erase::Erased<[u8; 8usize]>, rustc_query_system[2e8d7520d0881f27]::dep_graph::graph::DepNodeIndex>, false, false, false>, rustc_query_impl[23ea27e8a90d8b37]::plumbing::QueryCtxt, false>
  37:     0x7f09c4d99cc9 - rustc_query_impl[23ea27e8a90d8b37]::query_impl::mir_drops_elaborated_and_const_checked::get_query_non_incr::__rust_end_short_backtrace
  38:     0x7f09c4d96fbf - <rustc_middle[21423b4d0c52388d]::ty::context::TyCtxt>::par_hir_body_owners::<rustc_interface[eefada1bb51de1fc]::passes::run_required_analyses::{closure#1}::{closure#0}>::{closure#0}
  39:     0x7f09c4d946fa - rustc_interface[eefada1bb51de1fc]::passes::analysis
  40:     0x7f09c4d93b91 - rustc_query_impl[23ea27e8a90d8b37]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[23ea27e8a90d8b37]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[21423b4d0c52388d]::query::erase::Erased<[u8; 0usize]>>
  41:     0x7f09c5d80b20 - rustc_query_system[2e8d7520d0881f27]::query::plumbing::try_execute_query::<rustc_query_impl[23ea27e8a90d8b37]::DynamicConfig<rustc_query_system[2e8d7520d0881f27]::query::caches::SingleCache<rustc_middle[21423b4d0c52388d]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[23ea27e8a90d8b37]::plumbing::QueryCtxt, false>
  42:     0x7f09c5d80776 - rustc_query_impl[23ea27e8a90d8b37]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  43:     0x7f09c5f14b13 - <rustc_interface[eefada1bb51de1fc]::passes::create_and_enter_global_ctxt<core[f3ddf6de697c19ed]::option::Option<rustc_interface[eefada1bb51de1fc]::queries::Linker>, rustc_driver_impl[6aae568bf99046c3]::run_compiler::{closure#0}::{closure#2}>::{closure#2} as core[f3ddf6de697c19ed]::ops::function::FnOnce<(&rustc_session[b2db305ad29886f6]::session::Session, rustc_middle[21423b4d0c52388d]::ty::context::CurrentGcx, alloc[7f145662aca8c9c]::sync::Arc<rustc_data_structures[e7ed3fe572b1bf09]::jobserver::Proxy>, &std[4ba5a76045f2e409]::sync::once_lock::OnceLock<rustc_middle[21423b4d0c52388d]::ty::context::GlobalCtxt>, &rustc_data_structures[e7ed3fe572b1bf09]::sync::worker_local::WorkerLocal<rustc_middle[21423b4d0c52388d]::arena::Arena>, &rustc_data_structures[e7ed3fe572b1bf09]::sync::worker_local::WorkerLocal<rustc_hir[96c42a8d71e5a9b1]::Arena>, rustc_driver_impl[6aae568bf99046c3]::run_compiler::{closure#0}::{closure#2})>>::call_once::{shim:vtable#0}
  44:     0x7f09c5de1205 - rustc_interface[eefada1bb51de1fc]::interface::run_compiler::<(), rustc_driver_impl[6aae568bf99046c3]::run_compiler::{closure#0}>::{closure#1}
  45:     0x7f09c5d36b0a - std[4ba5a76045f2e409]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[eefada1bb51de1fc]::util::run_in_thread_with_globals<rustc_interface[eefada1bb51de1fc]::util::run_in_thread_pool_with_globals<rustc_interface[eefada1bb51de1fc]::interface::run_compiler<(), rustc_driver_impl[6aae568bf99046c3]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  46:     0x7f09c5d368dd - <std[4ba5a76045f2e409]::thread::lifecycle::spawn_unchecked<rustc_interface[eefada1bb51de1fc]::util::run_in_thread_with_globals<rustc_interface[eefada1bb51de1fc]::util::run_in_thread_pool_with_globals<rustc_interface[eefada1bb51de1fc]::interface::run_compiler<(), rustc_driver_impl[6aae568bf99046c3]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[f3ddf6de697c19ed]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  47:     0x7f09c5d34cb8 - <std[4ba5a76045f2e409]::sys::thread::unix::Thread>::new::thread_start
  48:     0x7f09bf6969cb - <unknown>
  49:     0x7f09bf71aa0c - <unknown>
  50:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/tmp/im/rustc-ice-2025-12-12T15_23_24-1521150.txt` to your bug report

note: compiler flags: -Z mir-enable-passes=+DataflowConstProp --crate-type lib

query stack during panic:
#0 [optimized_mir] optimizing MIR for `a::{closure#0}`
#1 [layout_of] computing layout of `{async fn body of a()}`
#2 [mir_drops_elaborated_and_const_checked] elaborating drops for `a`
#3 [analysis] running analysis passes on crate `BC3F301FE4B2C0AC50F0C3F13FAF1D2AE4AD05443021AFF4CB58B5260D844868`
end of query stack
error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0512`.

```

</p>
</details>

<!--
query stack:
#0 [optimized_mir] optimizing MIR for `a::{closure#0}`
#1 [layout_of] computing layout of `{async fn body of a()}`
#2 [mir_drops_elaborated_and_const_checked] elaborating drops for `a`
#3 [analysis] running analysis passes on crate `mvce`

-->

--- a/compiler/rustc_hir_typeck/src/intrinsicck.rs
+++ b/compiler/rustc_hir_typeck/src/intrinsicck.rs
@@ -8,6 +8,7 @@ use rustc_index::Idx;
 use rustc_middle::bug;
 use rustc_middle::ty::layout::{LayoutError, SizeSkeleton};
 use rustc_middle::ty::{self, Ty, TyCtxt, Unnormalized};
+use rustc_span::ErrorGuaranteed;
 use rustc_span::def_id::LocalDefId;
 use tracing::trace;
 
@@ -72,7 +73,7 @@ fn check_transmute<'tcx>(
     from: Ty<'tcx>,
     to: Ty<'tcx>,
     hir_id: HirId,
-) {
+) -> Result<(), ErrorGuaranteed> {
     let span = tcx.hir_span(hir_id);
     let normalize = |ty| {
         if let Ok(ty) = tcx.try_normalize_erasing_regions(typing_env, Unnormalized::new_wip(ty)) {
@@ -92,7 +93,7 @@ fn check_transmute<'tcx>(
 
     // Transmutes that are only changing lifetimes are always ok.
     if from == to {
-        return;
+        return Ok(());
     }
 
     let sk_from = SizeSkeleton::compute(from, tcx, typing_env, span);
@@ -104,7 +105,7 @@ fn check_transmute<'tcx>(
         && let Ok(sk_to) = sk_to
     {
         if sk_from.same_size(sk_to) {
-            return;
+            return Ok(());
         }
 
         // Special-case transmuting from `typeof(function)` and
@@ -119,7 +120,7 @@ fn check_transmute<'tcx>(
                 .with_note(format!("target type: {to}"))
                 .with_help("cast with `as` to a pointer instead")
                 .emit();
-            return;
+            return Ok(());
         }
     }
 
@@ -131,21 +132,25 @@ fn check_transmute<'tcx>(
     );
     if from == to {
         err.note(format!("`{from}` does not have a fixed size"));
-        err.emit();
+        Err(err.emit())
     } else {
         err.note(format!("source type: `{}` ({})", from, skeleton_string(from, sk_from)));
         err.note(format!("target type: `{}` ({})", to, skeleton_string(to, sk_to)));
-        err.emit();
+        Err(err.emit())
     }
 }
 
-pub(crate) fn check_transmutes(tcx: TyCtxt<'_>, owner: LocalDefId) {
+pub(crate) fn check_transmutes(tcx: TyCtxt<'_>, owner: LocalDefId) -> Result<(), ErrorGuaranteed> {
     assert!(!tcx.is_typeck_child(owner.to_def_id()));
     let typeck_results = tcx.typeck(owner);
-    let None = typeck_results.tainted_by_errors else { return };
+    if let Some(e) = typeck_results.tainted_by_errors {
+        return Err(e);
+    };
 
     let typing_env = ty::TypingEnv::post_analysis(tcx, owner);
+    let mut result = Ok(());
     for &(from, to, hir_id) in &typeck_results.transmutes_to_check {
-        check_transmute(tcx, typing_env, from, to, hir_id);
+        result = result.and(check_transmute(tcx, typing_env, from, to, hir_id));
     }
+    result
 }

--- a/compiler/rustc_middle/src/queries.rs
+++ b/compiler/rustc_middle/src/queries.rs
@@ -1118,7 +1118,7 @@ rustc_queries! {
     }
 
     /// Unsafety-check this `LocalDefId`.
-    query check_transmutes(key: LocalDefId) {
+    query check_transmutes(key: LocalDefId) -> Result<(), ErrorGuaranteed> {
         desc { "check transmute calls inside `{}`", tcx.def_path_str(key) }
     }
 

--- a/compiler/rustc_mir_transform/src/lib.rs
+++ b/compiler/rustc_mir_transform/src/lib.rs
@@ -541,11 +541,15 @@ fn mir_drops_elaborated_and_const_checked(tcx: TyCtxt<'_>, def: LocalDefId) -> &
         body.tainted_by_errors = Some(error_reported);
     }
 
+    let root = tcx.typeck_root_def_id_local(def);
+    if let Err(e) = tcx.check_transmutes(root) {
+        body.tainted_by_errors = Some(e);
+    }
+
     // Also taint the body if it's within a top-level item that is not well formed.
     //
     // We do this check here and not during `mir_promoted` because that may result
     // in borrowck cycles if WF requires looking into an opaque hidden type.
-    let root = tcx.typeck_root_def_id_local(def);
     match tcx.def_kind(root) {
         DefKind::Fn
         | DefKind::AssocFn

--- a/tests/rustdoc-ui/issues/issue-79494.rs
+++ b/tests/rustdoc-ui/issues/issue-79494.rs
@@ -1,4 +1,4 @@
 //@ only-64bit
 
 pub const ZST: &[u8] = unsafe { std::mem::transmute(1usize) };
-//~^ ERROR transmuting from 8-byte type to 16-byte type
+//~^ ERROR: cannot transmute between types of different sizes, or dependently-sized types [E0512]

--- a/tests/rustdoc-ui/issues/issue-79494.stderr
+++ b/tests/rustdoc-ui/issues/issue-79494.stderr
@@ -1,9 +1,12 @@
-error[E0080]: transmuting from 8-byte type to 16-byte type: `usize` -> `&[u8]`
+error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
   --> $DIR/issue-79494.rs:3:33
    |
 LL | pub const ZST: &[u8] = unsafe { std::mem::transmute(1usize) };
-   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `ZST` failed here
+   |                                 ^^^^^^^^^^^^^^^^^^^
+   |
+   = note: source type: `usize` (64 bits)
+   = note: target type: `&[u8]` (128 bits)
 
 error: aborting due to 1 previous error
 
-For more information about this error, try `rustc --explain E0080`.
+For more information about this error, try `rustc --explain E0512`.

--- a/tests/ui/consts/transmute-size-mismatch-before-typeck.rs
+++ b/tests/ui/consts/transmute-size-mismatch-before-typeck.rs
@@ -14,5 +14,4 @@ fn main() {
 }
 
 const ZST: &[u8] = unsafe { std::mem::transmute(1usize) };
-//~^ ERROR transmuting from
-//~| ERROR cannot transmute between types of different sizes
+//~^ ERROR cannot transmute between types of different sizes

--- a/tests/ui/consts/transmute-size-mismatch-before-typeck.stderr
+++ b/tests/ui/consts/transmute-size-mismatch-before-typeck.stderr
@@ -1,9 +1,3 @@
-error[E0080]: transmuting from word size type to 2 * word size type: `usize` -> `&[u8]`
-  --> $DIR/transmute-size-mismatch-before-typeck.rs:16:29
-   |
-LL | const ZST: &[u8] = unsafe { std::mem::transmute(1usize) };
-   |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^ evaluation of `ZST` failed here
-
 error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
   --> $DIR/transmute-size-mismatch-before-typeck.rs:16:29
    |
@@ -13,7 +7,6 @@ LL | const ZST: &[u8] = unsafe { std::mem::transmute(1usize) };
    = note: source type: `usize` (word size)
    = note: target type: `&[u8]` (2 * word size)
 
-error: aborting due to 2 previous errors
+error: aborting due to 1 previous error
 
-Some errors have detailed explanations: E0080, E0512.
-For more information about an error, try `rustc --explain E0080`.
+For more information about this error, try `rustc --explain E0512`.

--- a/tests/ui/layout/base-layout-is-sized-ice-123078.rs
+++ b/tests/ui/layout/base-layout-is-sized-ice-123078.rs
@@ -8,8 +8,7 @@ struct S {
 }
 
 const C: S = unsafe { std::mem::transmute(()) };
-//~^ ERROR the type `S` has an unknown layout
-//~| ERROR cannot transmute between types of different sizes, or dependently-sized types
+//~^ ERROR cannot transmute between types of different sizes, or dependently-sized types
 
 const _: [(); {
     C;

--- a/tests/ui/layout/base-layout-is-sized-ice-123078.stderr
+++ b/tests/ui/layout/base-layout-is-sized-ice-123078.stderr
@@ -16,12 +16,6 @@ help: the `Box` type always has a statically known size and allocates its conten
 LL |     a: Box<[u8]>,
    |        ++++    +
 
-error[E0080]: the type `S` has an unknown layout
-  --> $DIR/base-layout-is-sized-ice-123078.rs:10:1
-   |
-LL | const C: S = unsafe { std::mem::transmute(()) };
-   | ^^^^^^^^^^ evaluation of `C` failed here
-
 error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
   --> $DIR/base-layout-is-sized-ice-123078.rs:10:23
    |
@@ -31,7 +25,7 @@ LL | const C: S = unsafe { std::mem::transmute(()) };
    = note: source type: `()` (0 bits)
    = note: target type: `S` (the type `S` has an unknown layout)
 
-error: aborting due to 3 previous errors
+error: aborting due to 2 previous errors
 
-Some errors have detailed explanations: E0080, E0277, E0512.
-For more information about an error, try `rustc --explain E0080`.
+Some errors have detailed explanations: E0277, E0512.
+For more information about an error, try `rustc --explain E0277`.

--- a/tests/ui/transmute/transmute-in-async-fn.rs
+++ b/tests/ui/transmute/transmute-in-async-fn.rs
@@ -1,0 +1,12 @@
+//@compile-flags: -Zmir-enable-passes=+DataflowConstProp --crate-type lib
+//@ edition:2021
+pub async fn a() -> u32 {
+    unsafe { std::mem::transmute(1u64) }
+    //~^error: cannot transmute between types of different sizes, or dependently-sized types
+}
+
+pub async fn b() -> u32 {
+    let closure = || unsafe { std::mem::transmute(1u64) };
+    //~^ ERROR: cannot transmute between types of different sizes, or dependently-sized types [E0512]
+    closure()
+}

--- a/tests/ui/transmute/transmute-in-async-fn.stderr
+++ b/tests/ui/transmute/transmute-in-async-fn.stderr
@@ -1,0 +1,21 @@
+error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
+  --> $DIR/transmute-in-async-fn.rs:4:14
+   |
+LL |     unsafe { std::mem::transmute(1u64) }
+   |              ^^^^^^^^^^^^^^^^^^^
+   |
+   = note: source type: `u64` (64 bits)
+   = note: target type: `u32` (32 bits)
+
+error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
+  --> $DIR/transmute-in-async-fn.rs:9:31
+   |
+LL |     let closure = || unsafe { std::mem::transmute(1u64) };
+   |                               ^^^^^^^^^^^^^^^^^^^
+   |
+   = note: source type: `u64` (64 bits)
+   = note: target type: `u32` (32 bits)
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0512`.

